python/lsst/daf/butler/_butler.py

```diff
@@ -1597,6 +1599,161 @@ def retrieveArtifacts(
+        full_check : `bool`, optional
+            If `True`, an additional check will be made for dataset artifact
+            existence. This will involve additional overhead due to the need
+            to query an external system. If `False` registry and datastore
+            will solely be asked if they know about the dataset but no
+            check for the artifact will be performed.
+        collections : Any, optional
+            Collections to be searched, overriding ``self.collections``.
+            Can be any of the types supported by the ``collections`` argument
+            to butler construction.
+        **kwargs
+            Additional keyword arguments used to augment or construct a
+            `DataCoordinate`.  See `DataCoordinate.standardize`
+            parameters.
+
+        Returns
+        -------
+        existence : `DatasetExistence`
+            Object indicating whether the dataset is known to registry and
+            datastore. Evaluates to `True` if the dataset is present and known
+            to both.
+        """
+        existence = DatasetExistence.UNRECOGNIZED
+
+        if isinstance(dataset_ref_or_type, DatasetRef):
+            ref = dataset_ref_or_type
+            registry_ref = self.registry.getDataset(dataset_ref_or_type.id)
+            if registry_ref is not None:
+                existence |= DatasetExistence.RECORDED
+
+                if dataset_ref_or_type != registry_ref:
+                    # This could mean that storage classes differ, so we should
+                    # check for that but use the registry ref for the rest of
+                    # the method.
+                    if registry_ref.is_compatible_with(dataset_ref_or_type):
+                        # Use the registry version from now on.
+                        ref = registry_ref
+                    else:
+                        raise ValueError(
+                            f"The ref given to exists() ({ref}) has the same dataset ID as one "
+                            f"in registry but has different incompatible values ({registry_ref})."
+                        )
+        else:
+            try:
+                ref = self._findDatasetRef(dataset_ref_or_type, data_id, collections=collections, **kwargs)
+            except (LookupError, TypeError, NoDefaultCollectionError):
+                return existence
+            existence |= DatasetExistence.RECORDED
+
+        if self.datastore.knows(ref):
+            existence |= DatasetExistence.DATASTORE
+
+        if full_check:
+            if self.datastore.exists(ref):
+                existence |= DatasetExistence._ARTIFACT
+        elif existence != DatasetExistence.UNRECOGNIZED:
+            # Do not add this flag if we have no other idea about a dataset.
+            existence |= DatasetExistence._ASSUMED
+
+        return existence
+
+    def _exists_many(
+        self,
+        refs: Iterable[DatasetRef],
+        /,
+        *,
```

Comment:
Do you need `/, *,` if there is nothing in between?

Reply:
I'm not entirely certain. I was giving myself flexibility but this is a private API at the moment so we can do whatever here.